lib/fog/compute/ecloud.rb

```diff
@@ -0,0 +1,825 @@
+        end
+
+        def initialize(options = {})
+          @base_path               = options[:base_path] || "/cloudapi/ecloud"
+          @connections             = {}
+          @connection_options      = options[:connection_options] || {}
+          @host                    = options[:ecloud_host] || API_URL
+          @persistent              = options[:persistent] || false
+          @version                 = options[:ecloud_version] || "2013-06-01"
+          @authentication_method   = options[:ecloud_authentication_method] || :cloud_api_auth
+          @access_key              = options[:ecloud_access_key]
+          @private_key             = options[:ecloud_private_key]
+          if @private_key.nil? || @authentication_method == :basic_auth
+            @authentication_method = :basic_auth
+            @username              = options[:ecloud_username]
+            @password              = options[:ecloud_password]
+            if @username.nil? || @password.nil?
+              raise ArgumentError, "No credentials (cloud auth, or basic auth) passed!"
+            end
+          else
+            @hmac = Fog::HMAC.new("sha256", @private_key)
+          end
+        end
+
+        def request(params)
+          # Convert the uri to a URI if it's a string.
+          if params[:uri].is_a?(String)
+            params[:uri] = URI.parse(@host + params[:uri])
+          end
+          host_url = "#{params[:uri].scheme}://#{params[:uri].host}#{params[:uri].port ? ":#{params[:uri].port}" : ""}"
+
+          # Hash connections on the host_url ... There"s nothing to say we won"t get URI"s that go to
+          # different hosts.
+          @connections[host_url] ||= Fog::XML::Connection.new(host_url, @persistent, @connection_options)
+
+          # Set headers to an empty hash if none are set.
+          headers = set_extra_headers_for(params) || set_extra_headers_for({})
+
+          # Make the request
+          options = {
+            :expects => (params[:expects] || 200),
+            :method  => params[:method] || "GET",
+            :path    => params[:uri].path + "#{"?#{params[:uri].query}" if params[:uri].query}",
+            :headers => headers
+          }
+          unless params[:body].nil? || params[:body].empty?
+            options.merge!(:body => params[:body])
+          end
+          response = @connections[host_url].request(options)
+          # Parse the response body into a hash
+          unless response.body.empty?
+            if params[:parse]
+              document = Fog::ToHashDocument.new
+              parser = Nokogiri::XML::SAX::PushParser.new(document)
+              parser << response.body
+              parser.finish
+
+              response.body = document.body
+            end
+          end
+
+          response
+        end
+
+        private
+
+        # if Authorization and x-tmrk-authorization are used, the x-tmrk-authorization takes precendence.
+        def set_extra_headers_for(params)
```

Comment:
Assignment Branch Condition size for set_extra_headers_for is too high. [39.82/15]